File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool isPossible(int num, const vector<int> diffs, const vector<int> times, long long limit) {
    long long time = 0;
    for (int i = 0; i < diffs.size(); i++) {
        int d = diffs[i];
        if (d <= num) {
            time += times[i];
        } else {
            int time_cur = times[i];
            int time_prev = i == 0 ? 0 : times[i-1];
            long long t = (d - num) * (time_cur + time_prev) + time_cur;
            time += t;
        }
    }
    return time <= limit;
}

int solution(vector<int> diffs, vector<int> times, long long limit) {
    int answer = 0;
    long long left = 1, right = *max_element(diffs.begin(), diffs.end());
    while (left <= right) {
        long long mid = (left + right) / 2;
        
        if (isPossible(mid, diffs, times, limit)) {
            right = mid - 1;
            answer = mid;
        } else {
            left = mid + 1;
        }
    }
    return answer;
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
class Solution {
    public int solution(int[] diffs, int[] times, long limit) {
        int answer = 0;
        int left = 1;
        int right = 0;
        for (int d : diffs) {
            right = Math.max(d, right);
        }
        
        while (left <= right) {
            int mid = (left + right) / 2;
            
            if (possible(diffs, times, limit, mid)) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return answer;
    }
    
    private boolean possible(int[] diffs, int[] times, long limit, int level) {
        long total = 0;
        for (int i = 0; i < diffs.length; i++) {
            long time_cur = times[i];
            if (diffs[i] <= level) {
                total += time_cur;
            } else {
                long fail = diffs[i] - level;
                long time_prev = i > 0 ? times[i - 1] : 0;
                total += (time_cur + time_prev) * fail + time_cur;
            }
        }
        return total <= limit;
    }
}